ValkeyJSON.md

```diff
@@ -75,9 +75,15 @@ called KeyTable, which is a sharded hash table storing key tokens and reference
 
 ### Document Size Limit
 
-Having a limit on json key size is generally considered as a good design. Without the limit, a malicious program can
-repeatedly call "json.arrinsert" or similar commands to make a json key grow forever, leading to OOM. ValkeyJSON will 
-provide a module config - json.max-document-size - to set the size limit.
+We added a limit on JSON key size as a good design practice. Without the limit, a malicious program could repeatedly call 
+"json.arrinsert" or similar commands to make a JSON key grow indefinitely, potentially leading to an out-of-memory (OOM) condition.
+
+The ValkeyJSON module will provide a configuration option, json.max-document-size, that allows users to set a size limit for
+JSON key size.
+
+However, the default value of json.max-document-size will be set to -1, meaning the size of JSON key will be unlimited. 
+This decision is based on the observation that RedisJSON, a related project, does not have a built-in size limit, and 
```

Comment:
It's a bit contradictory to reference another project and set the value to unlimited but still introduce this configuration. 
How is this any different to string `APPEND` command? Don't we have the same risk over there?

Reply:
It's a compromised approach based on RFC review feedback. I still think json data type should have a bounded size. But there were objections during the JSON RFC review, saying Valkey core data type like hash doesn't have a size limit (although string value has a limit of 500MB). So, the argument is, for consistency with other data types, JSON shouldn't enforce size limit. So, we compromised. Json will have no size limit by default. If you want size limit, you can override json.max-document-size.

Reply:
Yeah, my point is why introduce it just for json documents. With this approach, we might end up with multiple such configs to bother about. I think we should think holistically for all data types and not individual cases separately. So, I feel alternative 2 mentioned below (removing the config) is a good way to start.

Reply:
I'm fine with removing the config outright.

Reply:
I think it's generally considered a mistake to not have the configs to limit object size. Especially with cluster mode, which can cause significant performance issues with large keys. I think it makes sense to keep the config with a default value of `infinite`. 

@hpatro Do we want to start an issue on the main repo about object limits in parallel though?

Reply:
Zero seems much more common than -1 to for unlimited/disabled. (Searching for the word "zero" in valkey.conf, I find the following examples of zero = unlimited: client-output-buffer-limit, tls-session-cache-size, auto-aof-rewrite-percentage, latency-monitor-threshold, stream-node-max-bytes. There are probably more, like the listpack size limits, that aren't even mentioned in the same way.)

Historically, there have been no limits to key sizes, I think the default should be unlimited. I don't mind adding configs though. Huge keys are a bit like KEYS or very heavy Lua scripts. These are also not blocked by default, though it's possible to limit using ACL.

If we add memory tracking to all types, can we set a single config to limit any the size of any value? I don't know if it makes sense.

@joehu21 There is no limit for string keys, but there is a limit for the RESP protocol strings. You can create a huge string using the APPEND command and retrieve parts of it using GETRANGE.

Reply:
Understood it's more common to use zero to represent unlimited. The intention of using -1 for unlimited is that we can use 0 as a means to disable write commands. If the limit is set to zero, all json write commands are rejected. But I am also fine with using zero to represent unlimited.

Reply:
Interesting point. I didn't think about this.

But it's also possible to make it read-only using ACL rules, right? Using zero for this seems like hack.

Reply:
>Historically, there have been no limits to key sizes, I think the default should be unlimited. I don't mind adding configs though. Huge keys are a bit like KEYS or very heavy Lua scripts. These are also not blocked by default, though it's possible to limit using ACL.

FYI, we still limit strings to the proto_max_bulk_len even with append. This codebase, https://github.com/valkey-io/valkey/blob/2df56d87c0ebe802f38e8922bb2ea1e4ca9cfa76/src/t_string.c#L40C12-L40C29, is called on the append flow. 

> Understood it's more common to use zero to represent unlimited. The intention of using -1 for unlimited is that we can use 0 as a means to disable write commands. If the limit is set to zero, all json write commands are rejected. But I am also fine with using zero to represent unlimited.

I have a minor preference for zero to indicate infinite. I think you can achieve the same functionality by just saying setting to 1 indicates disabled write commands. There are no valid json blobs that are 1 byte :) (ACL also works)

Reply:
agreed

Reply:
> FYI, we still limit strings to the proto_max_bulk_len even with append.

TIL

But DUMP can still return larger bulk replies than `proto_max_bulk_len`, right?

Reply:
> But DUMP can still return larger bulk replies than proto_max_bulk_len, right?

Yeah, we actually have gone in Amazon now that I look to omit this for slot migration, which is the only place that dump is ignored. It's on my checklist for whenever we finalize atomic slot migration to make sure it's incorporated.